File: src/irrgarten/Game.java
package irrgarten;

import java.util.ArrayList;

public class Game {
    static private final Integer MAX_ROUNDS = 10;
    private Integer currentPlayerIndex;
    private String log;
    private ArrayList<Monster> monsters = new ArrayList<Monster>();
    private Labyrinth labyrinth;
    private ArrayList<Player> players = new ArrayList<Player>();
    private Player currentPlayer;

    Game(Integer nplayers){
        this.currentPlayerIndex = Dice.whoStarts(nplayers);
        this.log = "";
        this.labyrinth = new Labyrinth(5,5,0,0);
        for(Integer i =0; i< nplayers;i++){
            this.players.add(new Player(Character.forDigit(i, 10),Dice.randomIntelligence(),Dice.randomStrength()));
        }
        this.currentPlayer = players.get(this.currentPlayerIndex);
        this.ConfigureLabyrinth();
        this.labyrinth.spreadPlayers(this.players);
        
    }
    public boolean finished(){
        return this.labyrinth.haveAWinner();
    }
    public boolean nextStep(Directions preferredDirection){
        this.log = "";
        boolean isdead = this.currentPlayer.dead();
        if(!isdead){
            Directions direction =this.actualDirection(preferredDirection);
            if(direction != preferredDirection){
                this.logPlayerNoOrders();
            }
            Monster monster = this.labyrinth.putPlayer(direction,currentPlayer);

            if(monster == null){
                this.logNoMonster();
            }
            else{
                this.manageReward(this.combat(monster));
            }
        }
        else{
            manageResurrection();
        }
        boolean endGame = this.finished();
        if(!endGame){
            this.nextPlayer();
        }
        return endGame;
    }
    public GameState getGameState(){
        String cadplayers ="";
        String cadmonsters = "";
        for(Integer i = 0; i< this.players.size(); i++){
            cadplayers+=players.get(i).toString() + "\n";
        }
        for(Integer i = 0; i< this.monsters.size(); i++){
            cadmonsters+=monsters.get(i).toString();
        }
        return new GameState(this.labyrinth.toString(), cadplayers,cadmonsters,this.currentPlayerIndex,finished(),log);
    }
    private void ConfigureLabyrinth(){
        this.labyrinth.addBlock(Orientation.VERTICAL, 2, 1,2);
        Monster cerbero = new Monster("cerbero", 10, 20);
        this.labyrinth.addMonster(1, 0, cerbero);
        this.monsters.add(cerbero);

    }
    private void nextPlayer(){
        if(this.currentPlayerIndex < players.size()-1){
            this.currentPlayerIndex+=1;
            this.currentPlayer = this.players.get(this.currentPlayerIndex);
        }
        else{
            this.currentPlayerIndex = 0;
            this.currentPlayer = this.players.get(0);
        }
    }
    private Directions actualDirection(Directions preferredDirections){
        int currentRow = currentPlayer.getRow();
        int currentCol = currentPlayer.getCol();
        ArrayList<Directions> validMoves = this.labyrinth.validMoves(currentRow, currentCol);
        return this.currentPlayer.move(preferredDirections,validMoves);
        //return this.currentPlayer.move(preferredDirections,this.labyrinth.validMoves(this.currentPlayer.getRow(), this.currentPlayer.getRow()));
    }

    private GameCharacter combat(Monster monster){
        float monsterAttack, playerAttack;
        int rounds =0;
        GameCharacter winner = GameCharacter.PLAYER;
        playerAttack = this.currentPlayer.attack();
        boolean lose = monster.defend(playerAttack);

        while ((!lose)&&(rounds < MAX_ROUNDS)){
            winner = GameCharacter.MONSTER;
            rounds+=1;
            monsterAttack = monster.attack();
            lose = this.currentPlayer.defend(monsterAttack);

            if(!lose){
                playerAttack = this.currentPlayer.attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack);

            }

        }
        
        this.logRounds(rounds, MAX_ROUNDS);
        return  winner;
    }
    private void manageResurrection(){
        boolean resurrect = Dice.resurrectPlayer();
        if(resurrect){
           // this.currentPlayer.resurrect();
            this.resurrectFuzzy();
            this.logResurrected();
        }
        else{
            this.logPlayerSkipTurn();
        }
    }
    private void manageReward(GameCharacter winner){
        if(winner == GameCharacter.PLAYER){
            this.currentPlayer.receiveReward();
            this.logPlayerWon();
        }
        else{
            this.logMonsterWon();
        }
    }
    private void resurrectFuzzy(){
        this.currentPlayer.resurrect();
        this.currentPlayer = new FuzzyPlayer(this.currentPlayer);
        this.players.set(this.currentPlayerIndex, this.currentPlayer);
    }
    private void logPlayerWon(){
        this.log+="Has ganado el combate! \n";
    }
    private void logMonsterWon(){
        this.log+="El monstruo ha ganadpo el combate! \n";
    }
    private void logResurrected(){
        this.log+="Acabas de resucitar\n";
    }
    private void logPlayerSkipTurn(){
        this.log+="Has perdido el turno por estar muerto \n";
    }
    private void logPlayerNoOrders(){
        this.log+="No ha sido posible seguir la instrucción\n";
    }
    private void logNoMonster(){
        this.log+="Te has movido a una celda vacia\n";
    }
    private void logRounds(Integer rounds , Integer max){

        this.log+=String.format("Se han jugado %d rondas de %d \n", rounds,max);    
    }

}

File: JUEGO-RUBY/irrgarten/Game.rb
require_relative 'GameCharacter'
require_relative 'Player'
require_relative 'Directions'
require_relative 'Monster'
require_relative 'Dice'
require_relative 'Orientation'
require_relative 'GameState'
require_relative 'Labyrinth'
module Irrgarten
    class Game
        @@MAX_ROUNDS = 10

        def initialize(nplayer)
            @current_player_index = Dice.who_starts(nplayer)
            @log = ""
            @labyrinth= Labyrinth.new(5,5,0,0)
            @monsters = []
            @players = Array.new(nplayer)
            for i in 0...nplayer 
                @players[i] = Player.new(i,Dice.random_intelligence,Dice.random_strength)
            end
            @current_player = @players[@current_player_index]
            self.configure_labyrinth
            @labyrinth.spread_players(@players)
        end
        def finished
            @labyrinth.have_a_winner
        end
        def next_step(prefferred_direction)
            @log = ""
            if(!@current_player.dead())
                direction = actual_direction(prefferred_direction)
                if(direction != prefferred_direction)
                    log_player_no_orders()
                end
                monster = @labyrinth.put_player(direction,@current_player)
                if(monster == nil)
                    log_no_monster()
                else
                    winner = combat(monster)
                    manage_reward(winner)
                end
            else
                manage_resurrection
            end
            end_game = self.finished
            if(!end_game)
                next_player
            end  
            return end_game
    end

        def game_state
            cplayers = ""
            for i in 0...@players.size
                cplayers.concat("#{@players[i].to_s}\n")
            end
            cmonsters = ""

            for i in 0...@monsters.size
                cmonsters.concat("#{@monsters[i].to_s}\n") 
            end

            return GameState.new(@labyrinth.to_s,cplayers,cmonsters,@current_player_index,self.finished,@log)
        end

        
        private 

        def configure_labyrinth
            @labyrinth.add_block(Orientation::HORIZONTAL,1,1,3)
            monster  = Monster.new('m1',1.5,5)
            monster2  = Monster.new('m2',4,10)
            monster3  = Monster.new('m3',2,3)
            @labyrinth.add_monster(2,2,monster)
            @labyrinth.add_monster(3,3,monster2)
            @labyrinth.add_monster(4,4,monster3)
            @monsters.push(monster)
            @monsters.push(monster2)
            @monsters.push(monster3)
            

        end
        def next_player 

            if @current_player_index == (@players.size - 1)
                @current_player_index =0
            else
                @current_player_index +=1
            end
            @current_player = @players[@current_player_index]   
        end
        def actual_direction(preferred_direction)
            row = @current_player.row
            col = @current_player.col
            directions = @labyrinth.valid_moves(row,col)
            @current_player.move(preferred_direction,directions)
        end 
        def manage_resurrection
            if (Dice::resurrect_player())
                @current_player.resurrect()
                log_resurrected()
            else 
                log_player_skip_turn()
            end
        end
        def manage_reward (winner)
            if winner == GameCharacter::PLAYER
                @current_player.recieve_reward()
                log_player_won
            else 
                log_monster_won
            end

        end

        def combat(monster)
            rounds = 0
            winner = GameCharacter::PLAYER
            player_attack = @current_player.attack()
            lose = monster.defend(player_attack)
            while (!lose && (rounds < @@MAX_ROUNDS))
                winner = GameCharacter::MONSTER
                rounds+=1
                monster_attack = monster.attack()
                lose = @current_player.defend(monster_attack)

                if (!lose)
                    player_attack = @current_player.attack()
                    winner = GameCharacter::PLAYER
                    lose = monster.defend(player_attack)

                end
                
            end 
            log_rounds(rounds, @@MAX_ROUNDS)
            return winner
        end

        def log_player_won
            @log.concat("¡#{@current_player.name} Ha ganado el combate!\n")
        end
        def log_monster_won
            @log.concat("!El monstruo ha ganado el combate¡\n")
        end
        def log_resurrected
            @log.concat("Acabas de resucitar\n")
        end
        def log_player_skip_turn
            @log.concat("Has perdido el turno por estar muerto\n")
        end
        def log_player_no_orders()
            @log.concat("No ha sido posible seguir las instrucciones\n")
        end
        def log_no_monster()
            @log.concat("Te has movido a una celda vacía \n")
        end
        def log_rounds(rounds ,max)
            @log += "Se han producido " + rounds.to_s + " rondas de " + max.to_s + "\n"
        end
    end
end
